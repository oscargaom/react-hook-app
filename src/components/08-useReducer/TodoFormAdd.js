import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoFormAdd = ({handleAddNewTodo}) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ""
    });
    
    // console.log(description);

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length < 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddNewTodo(newTodo);

        reset();
    };

    return (
        <>
            <h4> Agregar TODO </h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Agregar nueva tarea ..."
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
