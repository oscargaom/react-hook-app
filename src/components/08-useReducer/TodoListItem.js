import React from 'react'

export const TodoListItem = ({todo, index, handleToggle, handleRemoveOnClick}) => {

    const {id, done, desc} = todo;

    return (
        <li
            key={id}
            className="list-group-item"
        >
            <p 
                className={`${done && "complete"}`}
                onClick={() => handleToggle(id)}
            >
                {index + 1}. {desc}
            </p>

            <button
                name={id}
                className="btn-danger"
                onClick={() => { handleRemoveOnClick(id) }}
            >
                Eliminar
            </button>
        </li>
    )
}
