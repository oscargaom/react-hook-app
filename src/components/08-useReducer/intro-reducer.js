
// Se crea un arreglo con un objeto.
const initialState = [{
    id: 1,
    todo: "Solfeo",
    done: false
}];

const reducer = (state = initialState, action) => {

    if (action?.type === "agregar") {
        return [...state, action.payload];
    }

    return state;
};

const todos = reducer();

const newTodo = {
    id: 2,
    todo: "Practiar por 15 minutos",
    done: false
};

const agregarTodoAction = {
    type: "agregar",
    payload: newTodo
};

const newTodos = reducer(todos, agregarTodoAction);

console.log(todos);
console.log(newTodos);
