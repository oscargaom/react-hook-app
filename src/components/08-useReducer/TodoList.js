import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, handleToggle, handleRemoveOnlick}) => {
    return (
        <ul className="list-group list-group-flush">
        {
            todos.map((todo, idx) =>
                <TodoListItem 
                    key={todo.id}
                    todo={todo} 
                    index={idx} 
                    handleToggle={handleToggle} 
                    handleRemoveOnlick={handleRemoveOnlick}
                />
            )
        }
    </ul>
    )
}
