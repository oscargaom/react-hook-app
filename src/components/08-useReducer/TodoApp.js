import React, { useReducer, useEffect } from 'react'
import "../../styles/styles.css"
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoFormAdd } from './TodoFormAdd';

// const initialState = [{
//     id: new Date().getTime(),
//     desc: "Solfeo",
//     done: false
// }];

const init = () => {
    // return [{
    //     id: new Date().getTime(),
    //     desc: "Solfeo",
    //     done: false
    // }]
    return JSON.parse(localStorage.getItem("todos")) || [];
};


export const TodoApp = () => {

    // const [todos, dispatch] = useReducer(todoReducer, initialState);
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
        // console.log('entrando a useEffect');
        // console.log(todos);
    }, [todos]);

    const handleRemoveOnClick = (todoId) => {

        // const todosJson = JSON.parse(localStorage.getItem("todos"))
        // console.log(todosJson);
        // const { target: { name } } = e;
        // console.log(name);

        const action = {
            type: "remove",
            payload: todoId
        };

        dispatch(action);
    };

    const handleToggle = (todoId) => {
        console.log(todoId);
        dispatch({
            type: "toggle",
            payload: todoId
        });
    };

    const handleAddNewTodo = (todo) => {
        
        dispatch({
            type: "add",
            payload: todo
        });
    };

    return (
        <div>
            <h1>Todo App ({todos.length}) </h1>
            <hr />

            {/* <div className="row">
                <div className="col-7">
                    <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, idx) =>
                                (<li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p 
                                        className={`${todo.done && "complete"}`}
                                        onClick={() => handleToggle(todo.id)}
                                    >
                                        {idx + 1}. {todo.desc}
                                    </p>

                                    <button
                                        name={todo.id}
                                        className="btn-danger"
                                        onClick={() => { handleRemoveOnClick(todo.id) }}
                                    >
                                        Eliminar
                                    </button>
                                </li>)
                            )
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4> Agregar TODO </h4>
                    <hr />

                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="description"
                            className="form-control"
                            placeholder="Agregar nueva tarea ..."
                            autoComplete="off"
                            value={description}
                            onChange={handleInputChange}
                        />

                        <button
                            type="submit"
                            className="btn btn-outline-primary mt-1 btn-block"
                        >
                            Agregar
                        </button>
                    </form>
                </div>
            </div> */}
            <div className="row">
                <div className="col-7">
                    <TodoList
                        todos={todos}
                        handleToggle={handleToggle}
                        handleRemoveOnClick={handleRemoveOnClick}
                    />
                </div>
                <div className="col-5">
                    <TodoFormAdd 
                        handleAddNewTodo={handleAddNewTodo}
                    />
                </div>
            </div>
        </div>
    )
}
