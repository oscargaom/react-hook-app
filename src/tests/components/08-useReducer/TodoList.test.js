import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas del componente <TodoList />', () => {
    const handleToggle = jest.fn();
    const handleRemoveOnClick = jest.fn();
    
    const wrapper = shallow(<TodoList
                                todos={demoTodos}
                                handleToggle={handleToggle}
                                handleRemoveOnClick={handleRemoveOnClick}
                            />);

    test('Debe mostrarse el componente', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('Debe tener un número válido de componentes <TodoListItem /> ', () => {

        expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);
        
    });

    test('Debe tener las props validas', () => {
        
        expect(wrapper.find('TodoListItem').at(0).props().todo).toEqual(expect.any(Object));
        expect(wrapper.find('TodoListItem').at(0).prop('index')).toEqual(expect.any(Number));
        expect(wrapper.find('TodoListItem').at(0).props().handleToggle).toEqual(expect.any(Function));
        expect(wrapper.find('TodoListItem').at(0).prop('handleRemoveOnClick')).toEqual(expect.any(Function));

    });
    
})
