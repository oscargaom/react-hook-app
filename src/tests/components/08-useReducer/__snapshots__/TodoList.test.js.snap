// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <TodoList /> Debe mostrarse el componente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleRemoveOnClick={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Solfeo",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleRemoveOnClick={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Practice keyboards 5 mins",
        "done": false,
        "id": 2,
      }
    }
  />
  <TodoListItem
    handleRemoveOnClick={[MockFunction]}
    handleToggle={[MockFunction]}
    index={2}
    key="3"
    todo={
      Object {
        "desc": "MeditaciÃ³n",
        "done": false,
        "id": 3,
      }
    }
  />
</ul>
`;
